src/Microsoft.AspNet.Server.Kestrel/Http/PathNormalizer.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Text;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public static class PathNormalizer
+    {
+        public static string NormalizeToNFC(string path)
+        {
+            if (!path.IsNormalized(NormalizationForm.FormC))
+            {
+                path = path.Normalize(NormalizationForm.FormC);
+            }
+
+            return path;
+        }
+
+        public static string RemoveDotSegments(string path)
+        {
+            if (ContainsDotSegments(path))
+            {
+                var normalizedChars = ArrayPool<char>.Shared.Rent(path.Length);
+                var normalizedIndex = normalizedChars.Length;
+                var pathIndex = path.Length - 1;
+                var skipSegments = 0;
+
+                while (pathIndex >= 0)
+                {
+                    if (pathIndex >= 2 && path[pathIndex] == '.' && path[pathIndex - 1] == '.' && path[pathIndex - 2] == '/')
+                    {
+                        if (normalizedIndex == normalizedChars.Length || normalizedChars[normalizedIndex] != '/')
+                        {
+                            normalizedChars[--normalizedIndex] = '/';
+                        }
+
+                        skipSegments++;
+                        pathIndex -= 3;
+                    }
+                    else if (pathIndex >= 1 && path[pathIndex] == '.' && path[pathIndex - 1] == '/')
+                    {
+                        pathIndex -= 2;
+                    }
+                    else
+                    {
+                        while (pathIndex >= 0)
+                        {
+                            var lastChar = path[pathIndex];
+
+                            if (skipSegments == 0)
+                            {
+                                normalizedChars[--normalizedIndex] = lastChar;
+                            }
+
+                            pathIndex--;
+
+                            if (lastChar == '/')
+                            {
+                                break;
+                            }
+                        }
+
+                        if (skipSegments > 0)
+                        {
+                            skipSegments--;
+                        }
+                    }
+                }
+
+                path = new string(normalizedChars, normalizedIndex, normalizedChars.Length - normalizedIndex);
+                ArrayPool<char>.Shared.Return(normalizedChars);
+            }
+
+            return path;
+        }
+
+        private static bool ContainsDotSegments(string path)
```

Comment:
@cesarbs actually you don't need to back check  
@halter73 for Url Path Encoding `%2E` is a valid `.`, however, in utf8 bytes only `.` is `.` and only `/` is `/`

Good table at start of https://en.wikipedia.org/wiki/UTF-8#Description
> * Backward compatibility: One-byte codes are used only for the ASCII values 0 through 127. In this case the UTF-8 code has the same value as the ASCII code. The high-order bit of these codes is always 0. This means that ASCII text is valid UTF-8, and UTF-8 can be used for parsers expecting 8-bit extended ASCII even if they are not designed for UTF-8.

> * Clear distinction between multi-byte and single-byte characters: Code points larger than 127 are represented by multi-byte sequences, composed of a leading byte and one or more continuation bytes. The leading byte has two or more high-order 1s followed by a 0, while continuation bytes all have '10' in the high-order position.